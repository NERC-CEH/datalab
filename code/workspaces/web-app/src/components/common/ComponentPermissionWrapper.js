import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { SYSTEM_INSTANCE_ADMIN } from 'common/src/permissionTypes';

class ComponentWrapper extends Component {
  hasPermission() {
    const { userPermissions, permission } = this.props;
    return userPermissions.includes(permission) || userPermissions.includes(SYSTEM_INSTANCE_ADMIN);
  }

  render() {
    if (this.hasPermission()) {
      return (
        <div style={this.props.style} className={this.props.className}>
          {this.props.children}
        </div>
      );
    }

    return (null);
  }
}

ComponentWrapper.propTypes = {
  permission: PropTypes.string.isRequired,
  userPermissions: PropTypes.arrayOf(PropTypes.string).isRequired,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
  style: PropTypes.object,
};

export default ComponentWrapper;
