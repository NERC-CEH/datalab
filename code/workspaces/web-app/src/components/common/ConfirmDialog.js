import React from 'react';
import PropTypes from 'prop-types';
import Dialog from '@mui/material/Dialog';
import DialogTitle from '@mui/material/DialogTitle';
import DialogContent from '@mui/material/DialogContent';
import DialogActions from '@mui/material/DialogActions';
import DialogContentText from '@mui/material/DialogContentText';
import PrimaryActionButton from './buttons/PrimaryActionButton';
import SecondaryActionButton from './buttons/SecondaryActionButton';
import theme from '../../theme';

function ConfirmDialog({ state, onSubmit, title, body, onCancel }) {
  if (!state.open) return null;

  return (
    <Dialog open={state.open} maxWidth="md">
      <DialogTitle>{title}</DialogTitle>
      <DialogContent>
        <DialogContentText>{body}</DialogContentText>
      </DialogContent>
      <DialogActions>
        <PrimaryActionButton
          onClick={onSubmit}
        >
          Confirm
        </PrimaryActionButton>
        <SecondaryActionButton
          style={{ marginLeft: theme.spacing(1) }}
          onClick={onCancel}
        >
          Cancel
        </SecondaryActionButton>
      </DialogActions>
    </Dialog>
  );
}

ConfirmDialog.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  state: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired,
  body: PropTypes.string.isRequired,
};

export default ConfirmDialog;
