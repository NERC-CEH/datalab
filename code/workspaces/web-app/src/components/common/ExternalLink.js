import React from 'react';

const ExternalLink = ({ href, children }) => (
  <a href={href} target="_blank" rel="noopener noreferrer">
    {children}
  </a>
);

export default ExternalLink;
