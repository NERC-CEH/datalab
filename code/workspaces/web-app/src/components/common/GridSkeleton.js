import React from 'react';
import PropTypes from 'prop-types';
import Skeleton from '@material-ui/lab/Skeleton';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles((theme) => {
  const horizontalSpacing = theme.spacing(3);
  const verticalSpacing = theme.spacing(1);

  return {
    grid: {
      display: 'grid',
      gridTemplateColumns: ({ columns }) => `repeat(${columns}, 1fr)`,
      rowGap: `${verticalSpacing}px`,
      columnGap: `${horizontalSpacing}px`,
      margin: [[verticalSpacing, 0]],
    },
  };
});

const GridSkeleton = ({ rows = 3, columns = 3, sizingText = 'Text to infer size from', sizingTextVariant = 'body1', className }) => {
  const classes = useStyles({ columns });

  const contents = [];
  for (let i = 0; i < rows * columns; i++) {
    contents.push(
      <Skeleton key={i} variant="text">
        <Typography variant={sizingTextVariant}>{sizingText}</Typography>
      </Skeleton>,
    );
  }

  const finalClassName = constructClassName(classes.grid, className);
  return <div className={finalClassName}>{contents}</div>;
};

GridSkeleton.propTypes = {
  rows: PropTypes.number,
  columns: PropTypes.number,
  sizingText: PropTypes.string,
  sizingTextVariant: PropTypes.oneOf([ // The Material UI Typography variants
    'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
    'subtitle1', 'subtitle2',
    'body1', 'body2',
    'caption', 'button', 'overline', 'srOnly', 'inherit',
  ]),
  className: PropTypes.string,
};

const constructClassName = (...classNames) => (
  classNames
    .filter(name => name) // remove falsy values
    .join(' ')
);

export default GridSkeleton;
