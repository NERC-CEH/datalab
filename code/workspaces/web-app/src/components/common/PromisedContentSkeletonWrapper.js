import React from 'react';
import PropTypes from 'prop-types';

const PromisedContentSkeletonWrapper = ({ promises, skeletonComponent, skeletonProps = {}, children }) => {
  const fetching = Array.isArray(promises) ? promises.some(item => item.fetching) : promises.fetching;

  if (fetching) return React.createElement(skeletonComponent, skeletonProps);
  return children;
};

const promisePropTypes = PropTypes.shape({
  fetching: PropTypes.bool.isRequired,
});

PromisedContentSkeletonWrapper.propTypes = {
  promises: PropTypes.oneOfType([
    promisePropTypes.isRequired,
    PropTypes.arrayOf(promisePropTypes).isRequired,
  ]),
  skeletonComponent: PropTypes.oneOfType([
    PropTypes.node.isRequired,
    PropTypes.elementType.isRequired,
  ]),
  skeletonProps: PropTypes.object,
};

export default PromisedContentSkeletonWrapper;
