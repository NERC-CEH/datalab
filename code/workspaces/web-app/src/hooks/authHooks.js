import useShallowSelector from './useShallowSelector';
import authSelectors from '../selectors/authSelectors';

const { currentUserId, currentUserPermissions, currentUserTokens } = authSelectors;

export const useCurrentUserId = () => useShallowSelector(currentUserId);
export const useCurrentUserPermissions = () => useShallowSelector(currentUserPermissions);
export const useCurrentUserTokens = () => useShallowSelector(currentUserTokens);
