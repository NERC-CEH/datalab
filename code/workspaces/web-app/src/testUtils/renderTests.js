import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider as MuiThemeProvider } from '@mui/material/styles';
import theme from '../theme';

const AllTheProviders = ({ children }) => (
    <MuiThemeProvider theme={theme}>
        {children}
    </MuiThemeProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
