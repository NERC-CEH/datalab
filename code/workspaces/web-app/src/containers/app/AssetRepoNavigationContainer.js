import React from 'react';
import { Route, Switch, Redirect } from 'react-router-dom';
import { permissionTypes } from 'common';
import SideBarNavigation from '../../components/app/SideBarNavigation';
import AssetRepoAddMetadataPage from '../../pages/AssetRepoAddMetadataPage';
import AssetRepoFindPage from '../../pages/AssetRepoFindPage';
import AddAssetsToNotebookPage from '../../pages/AddAssetsToNotebookPage';
import RoutePermissions from '../../components/common/RoutePermissionWrapper';
import NotFoundPage from '../../pages/NotFoundPage';
import AssetRepoSideBar from '../../components/app/AssetRepoSideBar';

const { SYSTEM_DATA_MANAGER } = permissionTypes;

function AssetRepoNavigationContainer() {
  return (
    <SideBarNavigation sideBar={
      <AssetRepoSideBar />
    }>
      <Switch>
        <RoutePermissions
          exact
          path="/assets/add-metadata"
          component={AssetRepoAddMetadataPage}
          permission={SYSTEM_DATA_MANAGER}
        />
        <RoutePermissions
          exact
          path="/assets/find"
          component={AssetRepoFindPage}
          permission={SYSTEM_DATA_MANAGER}
        />
        <RoutePermissions
          exact
          path="/assets/add-to-notebook"
          component={AddAssetsToNotebookPage}
          permission={SYSTEM_DATA_MANAGER}
        />
        <Redirect exact from="/assets" to="/assets/add-metadata" />
        <Route component={NotFoundPage} />
      </Switch>
    </SideBarNavigation>
  );
}

export default AssetRepoNavigationContainer;
