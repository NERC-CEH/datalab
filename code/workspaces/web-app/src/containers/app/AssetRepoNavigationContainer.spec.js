import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { permissionTypes } from 'common';
import { render } from '../../testUtils/renderTests';
import AssetRepoNavigationContainer from './AssetRepoNavigationContainer';
import { useCurrentUserPermissions } from '../../hooks/authHooks';

jest.mock('../../pages/AssetRepoAddMetadataPage', () => () => (<>AssetRepoAddMetadataPage Mock</>));
jest.mock('../../pages/AssetRepoFindPage', () => () => (<>AssetRepoFindPage Mock</>));

jest.mock('../../components/app/Footer', () => () => (<div>Footer mock</div>));

jest.mock('../../hooks/authHooks');

describe('AssetRepoNavigationContainer', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    useCurrentUserPermissions.mockReturnValue({ value: [permissionTypes.SYSTEM_INSTANCE_ADMIN] });
  });

  const renderWithHistory = path => render(
    <MemoryRouter initialEntries={[path]} >
      <AssetRepoNavigationContainer />
    </MemoryRouter>,
  );

  it('renders to match snapshot passing correct props to children', () => {
    const wrapper = renderWithHistory('/foo/bar');
    expect(wrapper.container).toMatchSnapshot();

    expect(wrapper.queryByText('404 â€“ Page Not Found')).not.toBeNull();
  });

  it('renders correctly for the assets add metadata route', () => {
    const wrapper = renderWithHistory('/assets/add-metadata');

    expect(wrapper.queryByText('AssetRepoAddMetadataPage Mock')).not.toBeNull();
    expect(wrapper.queryByText('AssetRepoFindPage Mock')).toBeNull();
  });

  it('renders correctly for the assets find route', () => {
    const wrapper = renderWithHistory('/assets/find');

    expect(wrapper.queryByText('AssetRepoAddMetadataPage Mock')).toBeNull();
    expect(wrapper.queryByText('AssetRepoFindPage Mock')).not.toBeNull();
  });
});
