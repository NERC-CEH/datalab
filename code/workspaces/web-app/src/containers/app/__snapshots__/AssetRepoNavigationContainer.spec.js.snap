// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetRepoNavigationContainer renders to match snapshot passing correct props to children 1`] = `
<WithStyles(SideBarNavigation)
  sideBar={<ForwardRef(WithStyles) />}
>
  <Switch>
    <RoutePermissionWrapper
      component={[Function]}
      exact={true}
      path="/assets/add-metadata"
      permission="system:data:manager"
    />
    <RoutePermissionWrapper
      component={[Function]}
      exact={true}
      path="/assets/find"
      permission="system:data:manager"
    />
    <Redirect
      exact={true}
      from="/assets"
      to="/assets/add-metadata"
    />
    <Route
      component={[Function]}
    />
  </Switch>
</WithStyles(SideBarNavigation)>
`;
